File: cursed_time/lib/src/balatro/play.py
# balatro/play.py
from dataclasses import dataclass
from typing import List, Optional
from enum import Enum

# Import Card, Rank, and Suit from deck module
from .deck import Card, Rank, Suit

@dataclass
class HandCard:
    """Card in hand with selection state"""
    card: Optional[Card]
    selected: bool

# Result classes that match how they're used in smart_analyzer.py
class PlayResult:
    """Result of playing a hand"""
    AGAIN = "Again"
    ROUND_OVER = "RoundOver"
    GAME_OVER = "GameOver"
    
    def __init__(self, type_str: str, data=None, connection=None):
        self.type = type_str
        if type_str == self.AGAIN:
            self.play = Play(data, connection) if data else None
        elif type_str == self.ROUND_OVER:
            from .overview import RoundOverview
            self.overview = RoundOverview(data, connection) if data else None
        elif type_str == self.GAME_OVER:
            from .overview import GameOverview
            self.game_over = GameOverview(connection) if connection else None

class DiscardResult:
    """Result of discarding cards"""
    AGAIN = "Again"
    GAME_OVER = "GameOver"
    
    def __init__(self, type_str: str, data=None, connection=None):
        self.type = type_str
        if type_str == self.AGAIN:
            self.play = Play(data, connection) if data else None
        elif type_str == self.GAME_OVER:
            from .overview import GameOverview
            self.game_over = GameOverview(connection) if connection else None

class Play:
    """Play screen handler"""
    def __init__(self, play_info: dict, connection):
        self.info = play_info
        self.connection = connection
    
    def hand(self) -> List[HandCard]:
        """Get current hand"""
        hand = []
        for card_data in self.info["hand"]:
            if card_data["card"]:
                rank = Rank(card_data["card"]["rank"])
                suit = Suit(card_data["card"]["suit"])
                card = Card(rank=rank, suit=suit)
                hand.append(HandCard(card=card, selected=card_data["selected"]))
            else:
                hand.append(HandCard(card=None, selected=False))
        return hand
    
    def score(self) -> float:
        return self.info["score"]
    
    def hands(self) -> int:
        return self.info["hands"]
    
    def discards(self) -> int:
        return self.info["discards"]
    
    def money(self) -> int:
        return self.info["money"]
    
    def blind(self) -> dict:
        return self.info["current_blind"]
    
    async def click(self, indices: List[int]):
        """Click cards at specified indices"""
        response = await self.connection.send_request("play/click", {"indices": indices})
        
        if "Ok" in response["body"]:
            self.info = response["body"]["Ok"]
            return self
        else:
            raise Exception(f"Click failed: {response['body']['Err']}")
    
    async def play(self):
        """Play the selected cards"""
        response = await self.connection.send_request("play/play")
        
        if "Ok" in response["body"]:
            result_data = response["body"]["Ok"]
            
            if "Again" in result_data:
                return PlayResult(PlayResult.AGAIN, result_data["Again"], self.connection)
            elif "RoundOver" in result_data:
                return PlayResult(PlayResult.ROUND_OVER, result_data["RoundOver"], self.connection)
            elif "GameOver" in result_data:
                return PlayResult(PlayResult.GAME_OVER, None, self.connection)
        else:
            raise Exception(f"Play failed: {response['body']['Err']}")
    
    async def discard(self):
        """Discard the selected cards"""
        response = await self.connection.send_request("play/discard")
        
        if "Ok" in response["body"]:
            result_data = response["body"]["Ok"]
            
            if "Again" in result_data:
                return DiscardResult(DiscardResult.AGAIN, result_data["Again"], self.connection)
            elif "GameOver" in result_data:
                return DiscardResult(DiscardResult.GAME_OVER, None, self.connection)
        else:
            raise Exception(f"Discard failed: {response['body']['Err']}")

File: client/lib/src/balatro/play.rs
use serde::{Deserialize, Serialize};

use crate::net::Connection;
use super::{
    deck::Card,
    overview::{GameOverview, RoundOverview},
    Error,
    blinds::CurrentBlind
};

pub struct Play<'a> {
    info: protocol::PlayInfo,
    connection: &'a mut Connection,
}

impl<'a> Play<'a> {
    pub(crate) fn new(info: protocol::PlayInfo, connection: &'a mut Connection) -> Self {
        Self { info, connection }
    }

    pub fn blind(&self) -> &CurrentBlind {
        &self.info.current_blind
    }

    pub fn hand(&self) -> &[HandCard] {
        &self.info.hand
    }

    pub fn score(&self) -> &f64 {
        &self.info.score
    }

    pub fn hands(&self) -> &u8 {
        &self.info.hands
    }

    pub fn discards(&self) -> &u8 {
        &self.info.discards
    }

    pub fn money(&self) -> &u32 {
        &self.info.money
    }

    pub async fn click(self, indices: &[u32]) -> Result<Self, Error> {
        let info = self.connection.request(protocol::PlayClick { indices: indices.to_vec() }).await??;
        Ok(Self::new(info, self.connection))
    }

    pub async fn play(self) -> Result<PlayResult<'a>, Error> {
        let info = self.connection.request(protocol::PlayPlay).await??;
        let result = match info {
            protocol::PlayResult::Again(info) => PlayResult::Again(Self::new(info, self.connection)),
            protocol::PlayResult::RoundOver(info) => PlayResult::RoundOver(RoundOverview::new(info, self.connection)),
            protocol::PlayResult::GameOver(_) => PlayResult::GameOver(GameOverview::new(self.connection)),
        };
        Ok(result)
    }

    pub async fn discard(self) -> Result<DiscardResult<'a>, Error> {
        let info = self.connection.request(protocol::PlayDiscard).await??;
        let result = match info {
            protocol::DiscardResult::Again(info) => DiscardResult::Again(Self::new(info, self.connection)),
            protocol::DiscardResult::GameOver(_) => DiscardResult::GameOver(GameOverview::new(self.connection)),
        };
        Ok(result)
    }
}

pub enum PlayResult<'a> {
    Again(Play<'a>),
    RoundOver(RoundOverview<'a>),
    GameOver(GameOverview<'a>),
}

pub enum DiscardResult<'a> {
    Again(Play<'a>),
    GameOver(GameOverview<'a>),
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct HandCard {
    card: Option<Card>,
    selected: bool,
}

pub(crate) mod protocol {
    use serde::{Deserialize, Serialize};
    use crate::net::protocol::{Packet, Request, Response};
    use super::{HandCard, CurrentBlind};
    use crate::balatro::overview::protocol::RoundOverviewInfo;
    #[derive(Serialize, Deserialize, Clone)]
    pub struct PlayInfo {
        pub current_blind: CurrentBlind,
        pub hand: Vec<HandCard>,
        pub score: f64,
        pub hands: u8,
        pub discards: u8,
        pub money: u32
    }

    impl Response for PlayInfo {}

    impl Packet for PlayInfo {
        fn kind() -> String {
            "play/hand".to_string()
        }
    }

    #[derive(Serialize, Deserialize, Clone)]
    pub struct PlayClick {
        pub indices: Vec<u32>
    }

    impl Request for PlayClick {
        type Expect = Result<PlayInfo, String>;
    }

    impl Packet for PlayClick {
        fn kind() -> String {
            "play/click".to_string()
        }
    }

    #[derive(Serialize, Deserialize, Clone)]
    pub struct PlayPlay;

    impl Request for PlayPlay {
        type Expect = Result<PlayResult, String>;
    }

    impl Packet for PlayPlay {
        fn kind() -> String {
            "play/play".to_string()
        }
    }

    #[derive(Serialize, Deserialize, Clone)]
    pub struct PlayDiscard;

    impl Request for PlayDiscard {
        type Expect = Result<DiscardResult, String>;
    }

    impl Packet for PlayDiscard {
        fn kind() -> String {
            "play/discard".to_string()
        }
    }

    #[derive(Serialize, Deserialize, Clone)]
    pub enum PlayResult {
        Again(PlayInfo),
        RoundOver(RoundOverviewInfo),
        GameOver(Vec<()>),
    }

    impl Response for PlayResult {}

    impl Packet for PlayResult {
        fn kind() -> String {
            "play/play/result".to_string()
        }
    }

    #[derive(Serialize, Deserialize, Clone)]
    pub enum DiscardResult {
        Again(PlayInfo),
        GameOver(Vec<()>),
    }

    impl Response for DiscardResult {}

    impl Packet for DiscardResult {
        fn kind() -> String {
            "play/discard/result".to_string()
        }
    }
    
}
